### PR TITLE
[Attention] Flash Attention 3 - fp8

### PR DESCRIPTION
**Pull request by mickaelseznec**
This PR add support for FP8 KV cache with FlashAttention3 (related PR in flash-attn [here](https://github.com/vllm-project/flash-attention/pull/50)) cc @LucasWilkinson Please do not merge this PR as long as it's not referencing vllm-project/flash-attention yet.

FlashAttention (contrary to FlashInfer) does attention with all Q, K and V in FP8.
The performance is usually better than FlashInfer FP8 KV and FlashAttention 3 with bf16.

I added support for v0 and v1 + some unit testing.

Note that I've added a trick for checkpoints not providing q_scale and reuse the k_scale (with is something TRTLLM does fwiw).

Also: I added a small QoS improvement when debugging v1: workers send back their traceback when they raise an exception.

### CODE REVIEW

**Comment by mickaelseznec** [comment]
CI failing because` vllm/tests/entrypoints/openai/test_accuracy.py` from [here](https://github.com/vllm-project/vllm/blob/main/.buildkite/run-tpu-test.sh#L22) doesn't exist.

@robertgshaw2-redhat any idea how should I fix? Just rename in run-tpu-test.sh? (@NickLucche you moved the file)

**Comment by NickLucche** [comment]
This is a known issue, PR addressing it here https://github.com/vllm-project/vllm/pull/13898. It won't block your PR.

**Comment by NickLucche** [comment]
I see there's some other problem with building the image, but likely CI just needs another spin

**Comment by LucasWilkinson** [comment]
@mickaelseznec apologies for the delay, https://github.com/vllm-project/flash-attention/pull/50#pullrequestreview-2675615072 has been merged, you can now point to vllm_flash_attn

We will need to populate the sccache on the server to get it through the CI, I can help with this once the tag is updated :+1:

File: tests/kernels/test_flash_attn.py [hunk 1]
```diff
@@ -223,10 +261,30 @@ def test_varlen_with_paged_kv(
...
+        k_scale = key_cache.amax().to(torch.float32) / 448.0
+        v_scale = value_cache.amax().to(torch.float32) / 448.0
+
+        maybe_quantized_query = (query / q_scale).to(q_dtype)
+        maybe_quantized_key_cache = (key_cache / k_scale).to(q_dtype)
+        maybe_quantized_value_cache = (value_cache / k_scale).to(q_dtype)
+
+        q_descale = q_scale.expand((num_seqs, num_kv_heads))
+        k_descale = k_scale.expand((num_seqs, num_kv_heads))
+        v_descale = v_scale.expand((num_seqs, num_kv_heads))
```
Comment (LucasWilkinson) [inline tests/kernels/test_flash_attn.py:282]: nit: could we maybe test per-head scales here too?, i.e. also test with non-zero strides

File: vllm/platforms/cuda.py [hunk 2]
```diff
@@ -240,15 +240,6 @@ def get_attn_backend_cls(cls, selected_backend, head_size, dtype,
                 "Cannot use FlashAttention-2 backend for dtype other than "
                 "torch.float16 or torch.bfloat16.")
             target_backend = _Backend.XFORMERS
-        elif kv_cache_dtype is not None and \
```
Comment (LucasWilkinson) [inline vllm/platforms/cuda.py:243]: we should keep this check but restrict it to FA2, i.e. check `get_flash_attn_version() != 2` (`get_flash_attn_version()` is in `vllm/attention/backends/utils.py`)

**Review by LucasWilkinson — COMMENTED** [review]
Thanks for the contribution! Looks clean 😄, ill approve once we can get it updated to use `vllm_flash_attn`, added a couple comments

File: vllm/platforms/cuda.py (hunk 2 above)
Comment (mickaelseznec) [reply vllm/platforms/cuda.py]: Agree that this might be improved, but I can't directly import `get_flash_attn_version` because of circular dependency.

Do you prefer if I move that function in another file? `vllm/attention/backends/versions.py` for example?

Comment (LucasWilkinson) [reply vllm/platforms/cuda.py]: > Do you prefer if I move that function in another file? vllm/attention/backends/versions.py for example?

sure, maybe move it to:
```
vllm/attention/utils/fa_support.py
```
for now, since there is `is_flash_attn_mla_supported()` function that may come in https://github.com/vllm-project/vllm/pull/14258 so this could be a spot for both of those

Comment (mickaelseznec) [reply vllm/platforms/cuda.py:243]: I had to move it to `vllm/fa_utils.py` because of how `vllm/attention/__init__.py` imports a bunch of stuff for convenience.

**Comment by LucasWilkinson** [comment]
apologies for the delay, the CI should be working now. There appears to be failing kernel tests

**Review by LucasWilkinson — APPROVED** [review]

**Comment by JaheimLee** [comment]
Hi, it seems there is no new nightly wheel since this PR. Is there anything wrong? @LucasWilkinson
